modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/snapshot/SnapshotRestoreCacheGroupProcess.java

```diff
@@ -0,0 +1,647 @@
+
+        if (inProgress(null)) {
+            return new GridFinishedFuture<>(
+                new IgniteException(OP_REJECT_MSG + "The previous snapshot restore operation was not completed."));
+        }
+
+        DiscoveryDataClusterState state = ctx.state().clusterState();
+
+        if (state.state() != ClusterState.ACTIVE || state.transition())
+            return new GridFinishedFuture<>(new IgniteException(OP_REJECT_MSG + "The cluster should be active."));
+
+        // Skip creating future on initiator.
+        if (fut.isDone())
+            fut = new GridFutureAdapter<>();
+
+        opCtx = new SnapshotRestoreContext(req.requestId(), req.snapshotName(), req.nodes(), req.configs());
+
+        fut.listen(f -> opCtx = null);
+
+        if (!allNodesInBaselineAndAlive(req.nodes()))
+            return new GridFinishedFuture<>(new IgniteException(OP_REJECT_MSG + "Server node(s) has left the cluster."));
+
+        SnapshotRestoreContext opCtx0 = opCtx;
+
+        GridFutureAdapter<SnapshotRestoreEmptyResponse> retFut = new GridFutureAdapter<>();
+
+        try {
+            for (String grpName : opCtx0.groups())
+                ensureCacheAbsent(grpName);
+
+            for (StoredCacheData cfg : opCtx0.configs()) {
+                if (!F.isEmpty(cfg.config().getGroupName()))
+                    ensureCacheAbsent(cfg.config().getName());
+            }
+
+            if (!ctx.cache().context().snapshotMgr().snapshotLocalDir(opCtx0.snapshotName()).exists())
+                return new GridFinishedFuture<>();
+
+            boolean updateMeta = ctx.localNodeId().equals(req.updateMetaNodeId());
+
+            ctx.getSystemExecutorService().submit(() -> {
+                try {
+                    opCtx0.restore(updateMeta);
+
+                    if (!opCtx0.interrupted()) {
+                        retFut.onDone();
+
+                        return;
+                    }
+
+                    log.error("Snapshot restore process has been interrupted " +
+                        "[groups=" + opCtx0.groups() + ", snapshot=" + opCtx0.snapshotName() + ']', opCtx0.error());
+
+                    opCtx0.rollback();
+
+                    retFut.onDone(opCtx0.error());
+
+                }
+                catch (Throwable t) {
+                    retFut.onDone(t);
+                }
+            });
+
+            return retFut;
+        } catch (Exception e) {
+            return new GridFinishedFuture<>(e);
+        }
+    }
+
+    /**
+     * @param reqId Request ID.
+     * @param res Results.
+     * @param errs Errors.
+     */
+    private void finishPrepare(UUID reqId, Map<UUID, SnapshotRestoreEmptyResponse> res, Map<UUID, Exception> errs) {
+        GridFutureAdapter<Void> fut0 = fut;
+
+        if (fut0.isDone() || !reqId.equals(opCtx.requestId()))
+            return;
+
+        Exception failure = F.first(errs.values());
+
+        if (failure != null) {
+            opCtx.rollback();
```

Comment:
Deleting a large number of files may be a heavy operation.  We should use atomic operations like deleting the whole directory at once or perform such operations on executor service.